src/loader/gtfs/shape_prepare.cc

```diff
@@ -67,71 +72,237 @@ std::vector<shape_offset_t> get_offsets_by_dist_traveled(
   auto offsets = std::vector<shape_offset_t>{};
   offsets.reserve(dist_traveled_stops_times.size());
   auto remaining_shape_begin = begin(dist_traveled_shape_edges);
+  // Ensure final offset maps to valid shape point
+  auto const shape_end =
+      begin(dist_traveled_shape_edges) +
+      (end(dist_traveled_shape_edges) - begin(dist_traveled_shape_edges)) - 1;
   for (auto const& distance : dist_traveled_stops_times) {
-    remaining_shape_begin = std::lower_bound(
-        remaining_shape_begin, end(dist_traveled_shape_edges), distance);
+    remaining_shape_begin =
+        std::lower_bound(remaining_shape_begin, shape_end, distance);
     offsets.push_back(shape_offset_t{remaining_shape_begin -
                                      begin(dist_traveled_shape_edges)});
   }
   return offsets;
 }
 
-void calculate_shape_offsets(timetable const& tt,
-                             shapes_storage& shapes_data,
+shape_prepare::shape_results::result::result(
+    stop_seq_t const* stop_seq, std::vector<double> const* distances)
+    : stop_seq_{stop_seq},
+      distances_{distances},
+      offset_idx_{shape_offset_idx_t::invalid()},
+      boxes_{} {}
+
+shape_prepare::shape_prepare(shape_loader_state const& states,
                              vector_map<gtfs_trip_idx_t, trip> const& trips,
-                             shape_loader_state const& shape_states) {
+                             shapes_storage& shapes_data)
+    : index_offset_{states.index_offset_},
+      shape_results_(states.id_map_.size()),
+      shapes_{shapes_data},
+      results_ready_{false} {
+  for (auto i = 0U; i < shape_results_.size(); ++i) {
+    shape_results_[i].shape_idx_ = static_cast<shape_idx_t>(index_offset_ + i);
+  }
+  for (auto const& trip : trips) {
+    if (trip.shape_idx_ == shape_idx_t::invalid()) {
+      continue;
+    }
+    auto const idx = cista::to_idx(trip.shape_idx_ - index_offset_);
+    auto& results = shape_results_[idx].results_;
+    if (std::all_of(
+#if __cpp_lib_execution
+            std::execution::par_unseq,
```

Comment:
nicht alles mögliche wahrlos parallelisieren

keep it simple

premature optimization is the root of all evil

wenn's nicht signifikant im profiler auftaucht = egal => KISS